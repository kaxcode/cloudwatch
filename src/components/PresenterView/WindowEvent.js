import React from 'react';
import PropTypes from 'prop-types';

class WindowEvent extends React.PureComponent {
  static propTypes = {
    event: PropTypes.string.isRequired,
    handler: PropTypes.func.isRequired
  };

  componentDidMount() {
    this.addEvent();
  }

  componentWillUpdate() {
    this.removeEvent();
  }

  componentDidUpdate() {
    this.addEvent();
  }

  componentWillUnmount() {
    this.removeEvent();
  }

  render() {
    return null;
  }

  addEvent() {
    const { event, handler } = this.props;
    window.addEventListener(event, handler);
  }

  removeEvent() {
    const { event, handler } = this.props;
    window.removeEventListener(event, handler);
  }
}

export default WindowEvent;
