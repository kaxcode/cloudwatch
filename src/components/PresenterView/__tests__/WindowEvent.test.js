import React from 'react';
import Enzyme from 'enzyme';
import WindowEvent from '../WindowEvent';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });
jest.useFakeTimers();
const handleStorage = jest.fn();

describe('WindowEvent', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<WindowEvent event="storage" handler={handleStorage} />);
  });

  it('renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe('#componentWillUpdate', () => {
    const wrapper = shallow(
      <WindowEvent event="storage" handler={handleStorage} />
    );

    it('calls removeEvent', () => {
      const spy = jest.spyOn(wrapper.instance(), 'removeEvent');
      wrapper.instance().componentWillUpdate();
      expect(spy).toHaveBeenCalled();
    });
  });

  describe('#componentDidUpdate', () => {
    const wrapper = shallow(
      <WindowEvent event="storage" handler={handleStorage} />
    );

    it('calls addEvent', () => {
      const spy = jest.spyOn(wrapper.instance(), 'addEvent');
      wrapper.instance().componentDidUpdate();
      expect(spy).toHaveBeenCalled();
    });
  });

  describe('#componentWillUnmount', () => {
    const wrapper = shallow(
      <WindowEvent event="storage" handler={handleStorage} />
    );

    it('calls removeEvent', () => {
      const spy = jest.spyOn(wrapper.instance(), 'removeEvent');
      wrapper.instance().componentWillUnmount();
      expect(spy).toHaveBeenCalled();
    });
  });
});
