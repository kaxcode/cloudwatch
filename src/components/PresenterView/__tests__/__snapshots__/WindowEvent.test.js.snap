// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WindowEvent renders correctly 1`] = `""`;
